components/project-loader/project-loader.jsx

```diff
@@ -0,0 +1,87 @@
+import React from 'react';
+import Service from '../../js/service.js';
+import ProjectList from '../project-list/project-list.jsx';
+
+export default React.createClass({
+  getInitialState() {
+    return {
+      loadingData: false,
+      apiPageIndex: 1,
+      entries: [],
+      moreEntriesToFetch: false,
+      totalMatched: 0,
+      tokenOfExistingPromise: null
+    };
+  },
+  getDefaultProps() {
+    return {
+      params: {}
+    };
+  },
+  componentDidMount() {
+    this.fetchData();
+  },
+  componentWillReceiveProps(nextProps) {
+    // Reset state before fetching data for the new params.
+    // We want to keep existingPromise on record for fetchData
+    // to handle it accoringly.
+    let tokenOfExistingPromise = this.state.tokenOfExistingPromise;
+    let newState = Object.assign(this.getInitialState(), {tokenOfExistingPromise: tokenOfExistingPromise});
+
+    this.setState(newState, () => {
+      this.fetchData(nextProps.params);
+    });
+  },
+  fetchData(params = this.props.params) {
+    // Since we are making asynchronous calls to fetch data,
+    // we want to make sure exisiting fetch call is cancelled before we start a new one.
+    // This prevents any unfinished request to incorrectly update the component state
+    // when it returns.
+    if (this.state.tokenOfExistingPromise) {
+      this.state.tokenOfExistingPromise.cancel();
+    }
+
+    let combinedParams = Object.assign({}, params, { page: this.state.apiPageIndex });
+    let token = {};
+
+    this.setState({
+      // we are about to make a new request, set loadingData to true.
+      loadingData: true,
+      // we want component to know about the promise returned from the to-be-called Service.entries.get()
+      tokenOfExistingPromise: token
```

Comment:
Why are we putting this here, instead of on `this`? We don't use it for any UI rendering purposes, there is no need to have it live in state, especially as it needs to be modified outside of the React lifecycle functions.